test/e2e/api_inheritance/api_inheritance_test.go

```diff
@@ -0,0 +1,271 @@
+			return
+		}
+		crdAdminClient := apiExtensionsClients.Cluster("admin").ApiextensionsV1().CustomResourceDefinitions()
+
+		// Register CRDs in the admin logical cluster
+		requiredCrds := []metav1.GroupKind{
+			{Group: tenancy.GroupName, Kind: "workspaces"},
+		}
+
+		if err := config.BootstrapCustomResourceDefinitions(ctx, crdAdminClient, requiredCrds); err != nil {
+			t.Errorf("failed to bootstrap CRDs: %v", err)
+			return
+		}
+
+		kcpClients, err := clientset.NewClusterForConfig(cfg)
+		if err != nil {
+			t.Errorf("failed to construct kcp client for server: %v", err)
+			return
+		}
+
+		// For now, API inheritance only supports workspaces in the admin logical cluster, so let's create
+		// a couple there.
+		kcpAdminClient := kcpClients.Cluster("admin")
+
+		sourceWorkspace := &tenancyapi.Workspace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "source",
+			},
+		}
+		sourceWorkspace, err = kcpAdminClient.TenancyV1alpha1().Workspaces().Create(ctx, sourceWorkspace, metav1.CreateOptions{})
+		if err != nil {
+			t.Errorf("error creating source workspace: %v", err)
+			return
+		}
+
+		targetWorkspace := &tenancyapi.Workspace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "target",
+			},
+		}
+		targetWorkspace, err = kcpAdminClient.TenancyV1alpha1().Workspaces().Create(ctx, targetWorkspace, metav1.CreateOptions{})
+		if err != nil {
+			t.Errorf("error creating target workspace: %v", err)
+			return
+		}
+
+		// Install a CRD into source workspace
+		sourceCrdClient := apiExtensionsClients.Cluster("source").ApiextensionsV1().CustomResourceDefinitions()
+		if err := config.BootstrapCustomResourceDefinitions(ctx, sourceCrdClient, requiredCrds); err != nil {
+			t.Errorf("failed to bootstrap CRDs: %v", err)
+			return
+		}
+
+		// Make sure API group from CRD shows up in source workspace group discovery
+		if err := wait.PollImmediateUntilWithContext(ctx, 100*time.Millisecond, func(c context.Context) (done bool, err error) {
+			groups, err := kcpClients.Cluster("source").Discovery().ServerGroups()
+			if err != nil {
+				return false, fmt.Errorf("error retrieving source workspace group discovery: %w", err)
+			}
+			if groupExists(groups, tenancy.GroupName) {
+				return true, nil
+			}
+			return false, nil
+		}); err != nil {
+			t.Errorf("source workspace discovery is missing group %q", tenancy.GroupName)
+			return
+		}
+
+		// Make sure API resource from CRD shows up in source workspace group version discovery
+		resources, err := kcpClients.Cluster("source").Discovery().ServerResourcesForGroupVersion(tenancyapi.SchemeGroupVersion.String())
+		if err != nil {
+			t.Errorf("error retrieving source workspace tenancy API discovery: %v", err)
+			return
+		}
+		if !resourceExists(resources, "workspaces") {
+			t.Errorf("source workspace discovery is missing workspaces resource")
+			return
+		}
+
+		// This workspace will be created in the source workspace. This is to ensure that it doesn't
```

Comment:
super-nit: might be nice to use something other than Workspace as the type you're inheriting just to keep mind-melting levels down in this :)

Reply:
Switched to clusters